common/persistence/sql/sqlplugin/mysql/visibility.go

```diff
@@ -29,252 +29,186 @@ import (
 	row *sqlplugin.VisibilityRow,
-) (sql.Result, error) {
-	row.StartTime = mdb.converter.ToMySQLDateTime(row.StartTime)
-	row.ExecutionTime = mdb.converter.ToMySQLDateTime(row.ExecutionTime)
-	return mdb.conn.ExecContext(ctx,
-		templateCreateWorkflowExecutionStarted,
-		row.NamespaceID,
-		row.WorkflowID,
-		row.RunID,
-		row.StartTime,
-		row.ExecutionTime,
-		row.WorkflowTypeName,
-		row.Status,
-		row.Memo,
-		row.Encoding,
-		row.TaskQueue,
-	)
+) (result sql.Result, retError error) {
+	finalRow := mdb.prepareRowForDB(row)
+	tx, err := mdb.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := tx.Rollback()
+		// If the error is sql.ErrTxDone, it means the transaction already closed, so ignore error.
+		if err != nil && !errors.Is(err, sql.ErrTxDone) {
+			// Transaction rollback error should never happen, unless db connection was lost.
+			retError = fmt.Errorf("transaction rollback failed: %w", retError)
+		}
+	}()
+	result, err = tx.NamedExecContext(ctx, templateInsertWorkflowExecution, finalRow)
+	if err != nil {
+		return nil, err
+	}
+	_, err = tx.NamedExecContext(ctx, templateInsertCustomSearchAttributes, finalRow)
+	if err != nil {
+		return nil, err
+	}
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
 }
 
 // ReplaceIntoVisibility replaces an existing row if it exist or creates a new row in visibility table
 func (mdb *db) ReplaceIntoVisibility(
 	ctx context.Context,
 	row *sqlplugin.VisibilityRow,
-) (sql.Result, error) {
-	switch {
-	case row.CloseTime != nil && row.HistoryLength != nil:
-		row.StartTime = mdb.converter.ToMySQLDateTime(row.StartTime)
-		row.ExecutionTime = mdb.converter.ToMySQLDateTime(row.ExecutionTime)
-		closeTime := mdb.converter.ToMySQLDateTime(*row.CloseTime)
-		return mdb.conn.ExecContext(ctx,
-			templateCreateWorkflowExecutionClosed,
-			row.NamespaceID,
-			row.WorkflowID,
-			row.RunID,
-			row.StartTime,
-			row.ExecutionTime,
-			row.WorkflowTypeName,
-			closeTime,
-			row.Status,
-			*row.HistoryLength,
-			row.Memo,
-			row.Encoding,
-			row.TaskQueue,
-		)
-	default:
-		return nil, errCloseParams
+) (result sql.Result, retError error) {
+	finalRow := mdb.prepareRowForDB(row)
+	tx, err := mdb.db.BeginTxx(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := tx.Rollback()
+		// If the error is sql.ErrTxDone, it means the transaction already closed, so ignore error.
+		if err != nil && !errors.Is(err, sql.ErrTxDone) {
+			// Transaction rollback error should never happen, unless db connection was lost.
+			retError = fmt.Errorf("transaction rollback failed: %w", retError)
+		}
+	}()
+	result, err = tx.NamedExecContext(ctx, templateUpsertWorkflowExecution, finalRow)
+	if err != nil {
+		return nil, err
+	}
+	_, err = tx.NamedExecContext(ctx, templateUpsertCustomSearchAttributes, finalRow)
+	if err != nil {
+		return nil, err
 	}
+	err = tx.Commit()
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
```

Comment:
tbh it's hard for me to tell if this is a "correct" change, but if it's tested I believe it! Just flagging as something I don't fully understand. Perhaps the reasoning for some of the changes in this file should be explained in the description of the PR?

Reply:
So, Git is not show the changes in an ideal way. What I did was basically rename `visibility_v8.go` as `visibility.go`. No changes in logic.

Reply:
ah ok got it. I saw that for a few other files, just must not have read `visibility_v8.go` when reviewing this file